prov/efa/configure.m4

```diff
@@ -84,7 +84,16 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
 	      ])
 	CPPFLAGS=$save_CPPFLAGS
 
-	AS_IF([test $efa_happy -eq 1 ], [$1], [$2])
+	dnl Check for ibv_is_fork_initialized() in libibverbs
+	AS_IF([test $efa_happy -eq 1],
+		[AC_CHECK_DECL([ibv_is_fork_initialized],
+			[AC_DEFINE([HAVE_IBV_IS_FORK_INITIALIZED],
+				[1], [Define to 1 if libibverbs has ibv_is_fork_initialized])],
+			[],
+			[[#include <infiniband/verbs.h>]])
+		])
+
+	AS_IF([test $efa_happy -eq 1 -a x"$enable_efa" != x"no"], [$1], [$2])
```

Comment:
Why this line is moved to after the check and added `-a x"$enable_efa" != x"no"`?

Reply:
This went after all feature tests, I inserted the new test at the end of the feature tests, but before this. I added the extra check to ensure efa was enabled at Brian's request. The idea is that feature tests should run based on if efa_happy is 1, and then we should only run the passed arguments if all necessary features are present and efa is enabled.

HTH

Reply:
> I added the extra check to ensure efa was enabled at Brian's request. 

Thank you for doing that! Please do it via a separate commit.

> The idea is that feature tests should run based on if efa_happy is 1, and then we should only run the passed arguments if all necessary features are present and efa is enabled.

No sure I understand, to achieve that, should not you set `efa_happy ` before checking for ibv_is_fork_initialized.

Reply:
FWIW, it's a common pattern we use in Open MPI m4:

```
AS_IF([test $blah_happy -eq 1],
    [...do some test...
      set blah_happy=0 if the test fails])

AS_IF([test $blah_happy -eq 1],
    [...do some other test...
      set blah_happy=0 if the test fails])

AS_IF([test $blah_happy -eq 1],
    [...do yet another test...
      set blah_happy=0 if the test fails])

AS_IF([test $blah_happy -eq 1], [$1], [$2])
```

This lets you stop running tests if one of them fails.

So you might even want to change your code to:

```
AS_IF([test "$enable_efa" = "no"], [efa_happy=0])
```

Then your final condition can still be the same:

```
AS_IF([test $efa_happy -eq 1], [...])
```

But we're getting in to toe-MAY-toe vs. toe-MAH-toe here.  I.e., either one is fine; what I outlined above is what I personally prefer, but there's not a huge technical reason to go one way or the other.

Reply:
ibv_is_fork_initialized is not required for EFA, so the presence does not affect efa_happy

Reply:
But in the commit, you first check `test $efa_happy -eq 1`, then go ahead check for `ibv_is_fork_initialized`, so may be remove that `AS_IF`.

Reply:
> ibv_is_fork_initialized is not required for EFA, so the presence does not affect efa_happy

Understood.  My example showed folding `$enable_efa` into `$efa_happy`.

Reply:
> But we're getting in to toe-MAY-toe vs. toe-MAH-toe here. I.e., either one is fine; what I outlined above is what I personally prefer, but there's not a huge technical reason to go one way or the other.

I would much prefer the suggested approach actually, it is much more clear what is going on compared to `test foo -a bar`

Reply:
I asked Peter to only run the check for `ibv_is_fork_initialized` if efa_happy is 1.  That's because if efa_happy is 0, then running further tests is unnecessary and slows down configure for the end user.  Worse, you can cause side-effects if you have tests that rely on libraries that are causing failures (not the case here, but happens often).  The pattern Jeff called out above is the right one, we're just not consistent in following it in the EFA provider these days.